### PR TITLE
Make `dyn` a keyword in the 2018 edition

### PR DESCRIPTION
**Pull request by varkor**
Proposed in https://github.com/rust-lang/rust/issues/44662#issuecomment-421596088.

### CODE REVIEW

File: src/test/ui/rust-2018/dyn-keyword.rs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+// edition:2015
+// run-rustfix
+
+#![allow(unused_variables)]
+#![deny(keyword_idents)]
+
+fn main() {
+    let dyn = (); //~ ERROR dyn
+    //~^ WARN hard error in the 2018 edition
+}
```
Comment (Centril) [inline src/test/ui/rust-2018/dyn-keyword.rs]: + You'll want tests that this is correctly rustfix-ed.
+ Also, a test that this isn't a proper keyword on Rust 2015 is needed.

Comment (varkor) [reply src/test/ui/rust-2018/dyn-keyword.rs]: I forgot to commit the `.stderr` and `.fixed` files 🙄
I think this test should cover not being a proper keyword on Rust 2015, as the error message specifically pertains to the 2018 edition.

File: src/libsyntax_pos/symbol.rs [hunk 2]
```diff
@@ -414,26 +414,26 @@ declare_keywords! {
...
 
     // Weak keywords, have special meaning only in specific contexts.
-    (55, Auto,               "auto")
-    (56, Catch,              "catch")
-    (57, Default,            "default")
-    (58, Dyn,                "dyn")
+    (56, Auto,               "auto")
+    (57, Catch,              "catch")
+    (58, Default,            "default")
+
```
Comment (petrochenkov) [inline src/libsyntax_pos/symbol.rs]: Nit: stray newline

**Comment by petrochenkov** [comment]
You may also want to adjust the use of `keywords::Dyn` in `fn parse_ty_common` and always treat it as a start of type if `self.span.edition() == Edition2018`, this will avoid the fallback error `"expected type, found dyn"`

**Comment by petrochenkov** [comment]
`dyn` also needs to be explicitly listed in `fn ident_can_begin_type` now.

I would also be nice to have a copy of `test/ui/dyn-trait-compatibility.rs` compiled on 2018 edition.

**Comment by petrochenkov** [comment]
Blocked on FCP in https://github.com/rust-lang/rust/issues/44662#issuecomment-421596088

**Comment by Centril** [comment]
@petrochenkov, @varkor I think we can go ahead and merge this (assuming the implementation is correct) since all but 2 lang team members checked their boxes on the poll and those members were OK with this on the meeting.

**Pull request merged**